### PR TITLE
repo-updater: soft deletes for repos in DB store

### PR DESCRIPTION
**Pull request by keegancsmith**
This adds soft deletes to our DB store. The syncer has no knowledge of soft
deletions, instead the store will update soft deleted rows if they have the
same external ID.

When soft deleting a repository we give it a unique name. This is to prevent
potential future conflicts on name. Alternatively we could remove the unique
constraint on name, but this is a much larger change. Alternatively we could
make the constraint partial on deleted_at being null. Either of these could be
a follow up PR, or this solution can be seen as sufficient.

This commit updates the existing tests check invariants we expect on soft
deletions:

1. When undoing a soft-delete we get back the same ID we had before
2. We don't get conflicts on name against soft deleted repos.

Fixes https://github.com/sourcegraph/sourcegraph/issues/6859

### CODE REVIEW

**Comment by keegancsmith** [comment]
Two options "undeleting":
1. undeleted Repo appears in Modified LOSER
   - This requires changing NewDiff
   - already soft-deleted repos appear in unmodified
   - only repos we have never seen before appear in Added
2. undeleted Repo appears in Added WINNER
   - this requires changing DB Upsert
   - insert logic needs to handle conflicts / become an upsert
   - existing semantics around what a diff is remains the same

There is tech debt around external service being nullable that is hindering progress here.

File: cmd/repo-updater/repos/store.go [hunk 1]
```diff
@@ -675,10 +675,13 @@ ORDER BY batch.ordinality
 `
 
 var deleteReposQuery = batchReposQueryFmtstr + `
-DELETE FROM repo USING batch
+UPDATE repo
+SET
+  name = 'DELETED-' || extract(epoch from transaction_timestamp()) || '-' || batch.name,
```
Comment (keegancsmith) [inline cmd/repo-updater/repos/store.go]: :'( sorry

**Comment by keegancsmith** [comment]
@unknwon this seems to be working, but I have just got the tests to pass and its the end of my work day. I'll validate it more tomorrow and potentially submit as a real PR.

**Comment by ryanslade** [comment]
> When soft deleting a repository we give it a unique name. This is to prevent
> potential future conflicts on name. Alternatively we could remove the unique
> constraint on name, but this is a much larger change. Alternatively we could
> make the constraint partial on deleted_at being null. Either of these could be
> a follow up PR, or this solution can be seen as sufficient.

Without a  lot of background knowledge making the constraint partial sounds like the best
approach to me.

File: cmd/repo-updater/repos/store_test.go [hunk 2]
```diff
@@ -647,10 +651,52 @@ func testStoreUpsertRepos(store repos.Store) func(*testing.T) {
...
+				r.ID = 0
+				r.ExternalRepo.ID += "-different"
+			}
+			if err = tx.UpsertRepos(ctx, deleted...); err != nil {
+				t.Fatalf("UpsertRepos deleted error: %s", err)
+			} else if err = tx.UpsertRepos(ctx, want...); err != nil {
+				t.Fatalf("UpsertRepos want error: %s", err)
+			} else if have, err = tx.ListRepos(ctx, repos.StoreListReposArgs{}); err != nil {
+				t.Errorf("ListRepos error: %s", err)
+			} else if diff := pretty.Compare(have, want); diff != "" {
```
Comment (unknwon) [inline cmd/repo-updater/repos/store_test.go]: TIL: `pretty.Compare`

Comment (keegancsmith) [reply cmd/repo-updater/repos/store_test.go]: FYI We use go-cmp in other places. Go-cmp is better, just wanted to be consistent with the rest of the file. Go-cmp produces more succinct diffs as well as having a better version of reflect.DeepEqual

Comment (unknwon) [reply cmd/repo-updater/repos/store_test.go]: Good to know, thanks!

**Pull request merged**